File: android/src/main/java/com/plugins/forgerockbridge/ForgerockBridgePlugin.java
package com.plugins.forgerockbridge;

import android.content.Context;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.plugins.forgerockbridge.enums.ForgotPasswordEnum;
import com.plugins.forgerockbridge.nodeListenerCallbacks.ForgotPasswordNodeListener;
import com.plugins.forgerockbridge.state.PluginState;
import com.plugins.forgerockbridge.nodeListenerCallbacks.OTPNodeListener;
import com.plugins.forgerockbridge.nodeListenerCallbacks.OTPDeleteNodeListener;

@CapacitorPlugin(name = "ForgerockBridge")
public class ForgerockBridgePlugin extends Plugin {

    public static Context context;
    private final PluginState pluginState = new PluginState();


    @Override
    public void load() {
        super.load();
        context = getContext();
    }

    @PluginMethod
    public void initialize(PluginCall call) {
        SDKInitializer.handle(call, context);
    }

    @PluginMethod
    public void authenticate(PluginCall call) {
        AuthenticationHandler.handle(call, getContext(), pluginState);
    }

    @PluginMethod
    public void userInfo(PluginCall call) {
        UserHandler.getUserInfo(call);
    }

    @PluginMethod
    public void logout(PluginCall call) {
        UserHandler.logout(call);
    }

    @PluginMethod
    public void initializeOTPRegister(PluginCall call) {
        OTPTokenHandler.startOtpJourney(call, context, new OTPNodeListener(call, context, pluginState));
    }

    @PluginMethod
    public void deleteOTPRegister(PluginCall call) {
        OTPTokenHandler.startOtpJourney(call, context, new OTPDeleteNodeListener(call, context, pluginState));
    }

    @PluginMethod
    public void hasRegisteredMechanism(PluginCall call) {
        OTPTokenHandler.hasRegisteredMechanism(call, context);
    }

    @PluginMethod
    public void validateOTP(PluginCall call) {
        OTPTokenHandler.validateExistenceOTP(call, context);
    }

    @PluginMethod
    public void generateOTP(PluginCall call) {
        OTPTokenHandler.generateOTP(call, context);
    }

    @PluginMethod
    public void initForgotPassword(PluginCall call) {
        pluginState.reset();
        ForgotPasswordHandler.startForgotPasswordJourney(call, context, new ForgotPasswordNodeListener(call, context, pluginState, ForgotPasswordEnum.IdPath.INIT_FORGOT_PASS));
    }

    @PluginMethod
    public void getQuestionForgotPassword(PluginCall call) {
        ForgotPasswordHandler.getSecurityQuestion(call, pluginState);
    }

    @PluginMethod
    public void answerQuestionForgotPassword(PluginCall call) {
        ForgotPasswordHandler.answerQuestionForgotPassword(
            call,
            pluginState,
            context
        );
    }

    @PluginMethod
    public void changePasswordForgotPassword(PluginCall call) {
        ForgotPasswordHandler.changePasswordForgotPassword(
                call,
                pluginState,
                context
        );
    }

}

File: ios/Sources/ForgerockBridgePlugin/ForgerockBridgePlugin.swift
import Foundation
import FRAuth
import Capacitor


/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(ForgerockBridgePlugin)
public class ForgerockBridgePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ForgerockBridgePlugin"
    public let jsName = "ForgerockBridge"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "authenticate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "logout", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "userInfo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getAccessToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "initializeOTPRegister", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "validateOTP", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "deleteOTPRegister", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "generateOTP", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hasRegisteredMechanism", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "initForgotPassword", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getQuestionForgotPassword", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "answerQuestionForgotPassword", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "changePasswordForgotPassword", returnType: CAPPluginReturnPromise),
    ]
    public var pendingNode: Node? = nil
    public var didSubmitConfirmation = false

    /// Inicializa el SDK de ForgeRock con la configuración enviada desde el frontend.
    /// Parámetros requeridos: url, realm, journey, oauthClientId, oauthScope.
    @objc func initialize(_ call: CAPPluginCall) {
        SDKInitializer.initialize(call: call)
    }

    /// Inicia o continúa el flujo de autenticación. Maneja intentos, errores y callbacks como ConfirmationCallback.
    @objc func authenticate(_ call: CAPPluginCall) {
        AuthenticationHandler(call: call, plugin: self).authenticate()
    }

    /// Cierra la sesión del usuario actual. Limpia los tokens y estado del SDK.
    @objc func logout(_ call: CAPPluginCall) {
        UserHandler.logout(call: call)
    }

    /// Obtiene la información del usuario autenticado actual (claims del ID Token).
    @objc func userInfo(_ call: CAPPluginCall) {
        UserHandler.userInfo(call: call)
    }

    /// Retorna el Access Token del usuario actual para autenticación en APIs externas.
    @objc func getAccessToken(_ call: CAPPluginCall) {
        TokenHandler.getAccessToken(call: call)
    }
    
    // Inicializa el metodo de token y registro 
    @objc func initializeOTPRegister(_ call: CAPPluginCall) {
        let handler = OTPTokenHandler(call: call, plugin: self)
        let cb = NodeOTPCallBacks(call: call, plugin: self)
        handler.startJourney(call, completion: cb.handle)

    }

     // Eliminar token registrado 
    @objc func deleteOTPRegister(_ call: CAPPluginCall) {
        let handler = OTPTokenHandler(call: call, plugin: self)
        let cb = NodeOTPDeleteCallBacks(call: call, plugin: self)
        handler.startJourney(call, completion: cb.handle)
    }

    @objc func validateOTP(_ call: CAPPluginCall) {
        let handler = OTPTokenHandler(call: call, plugin: self)
        handler.validateExistenceOTP(call: call);
    }

    @objc func hasRegisteredMechanism(_ call: CAPPluginCall) {
        let handler = OTPTokenHandler(call: call, plugin: self)
        handler.hasRegisteredMechanism(call);
    }

    @objc func generateOTP(_ call: CAPPluginCall) {
        let handler = OTPTokenHandler(call: call, plugin: self)
        handler.generateOTP(call: call);
    }
    
     @objc func initForgotPassword(_ call: CAPPluginCall) {
        let handler = ForgotPasswordHandler(call: call, plugin: self)
        let cb = NodeForgotPasswordCallbacks(call: call, plugin: self)
        handler.startForgotPasswordJourney(call, completion: cb.handle);
    }

    @objc func getQuestionForgotPassword(_ call: CAPPluginCall) {
        let handler = ForgotPasswordHandler(call: call, plugin: self)
        handler.getSecurityQuestion(call);
    }

    @objc func answerQuestionForgotPassword(_ call: CAPPluginCall) {
        let handler = ForgotPasswordHandler(call: call, plugin: self)
        let cb = NodeForgotPasswordCallbacks(call: call, plugin: self)
        handler.answerQuestionForgotPassword(call, completion: cb.handle);
    }

    @objc func changePasswordForgotPassword(_ call: CAPPluginCall) {
        let handler = ForgotPasswordHandler(call: call, plugin: self)
        let cb = NodeForgotPasswordCallbacks(call: call, plugin: self)
        handler.changePasswordForgotPassword(call, completion: cb.handle);
    }
}
